core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala

```diff
@@ -585,17 +591,25 @@ class ExternalAppendOnlyMap[K, V, C](
       } else {
         logInfo(s"Task ${context.taskAttemptId} force spilling in-memory map to disk and " +
           s"it will release ${org.apache.spark.util.Utils.bytesToString(getUsed())} memory")
-        nextUpstream = spillMemoryIteratorToDisk(upstream)
+        val nextUpstream = spillMemoryIteratorToDisk(upstream)
+        assert(!upstream.hasNext)
         hasSpilled = true
+        upstream = nextUpstream
         true
       }
     }
 
+    private def destroy() : Unit = {
+      freeCurrentMap()
+      upstream = Iterator.empty
```

Comment:
Why `empy`, not `null`?

Reply:
Safer, class remains usable if for some reason hasNext is called again, and this costs absolutely nothing.

Reply:
+1